lib/remap-lang/src/object.rs

```diff
@@ -38,22 +68,22 @@ pub trait Object: std::fmt::Debug {
     /// When inserting into a coalesced path, the implementor is encouraged to
     /// insert into the right-most segment if none exists, but can return an
     /// error if needed.
-    fn insert(&mut self, path: &Path, value: Value) -> Result<(), String>;
+    fn insert(&mut self, path: &Path, value: Value) -> Result<(), Error>;
```

Comment:
I don't think exposing this type in this trait is the right approach to take.

I prefer we keep the error type of these functions free-form, such that the implementer can decide if they want to return a "structured error" or just some text.

What we _could_ do (and maybe should), is that we add an associated type to the trait:

```rust
 pub trait Object: std::fmt::Debug {
	type Err: std::fmt::Display;
}
```

This allows the implementer to use structured errors if they want to (such as is the case with the `impl Object for MetricEvent`), or use an unstructured `String`.

Basically, what I'm suggesting is we revert 733c44511509575895ec5a5094aa51c374d87a27, then move the structured errors into Vector itself, and have it set `type Err = Error`.

In remap we then just print the final string from the resulting error type we get back.

Reply:
However, we might as well _not_ add the `Err` associated type, and just have them return a `String` (as it was before), as the callee can always call `to_string()` themselves before they pass their structured error to TRL.